File: successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int[] answer = new int[spells.length];

        int m = potions.length;
        int maxPotion = potions[m-1];

        for (int i = 0; i < spells.length; i++) {
            int spell = spells[i];

            long minPotion = (long) Math.ceil(success * 1.0/spell);

            if (minPotion > maxPotion) {
                answer[i] = 0;
                continue;
            }
            int index = binarySearch(potions, (int) minPotion);
            answer[i] = m - index;
        }

        return answer;
    }

    private int binarySearch(int[] arr, int key) {
        int low = 0;
        int high = arr.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] < key) {
                low = mid + 1;
            }
            else {
                high = mid - 1;
            }
        }
        return low;
    }
}

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        result = []
        for spell in spells:
            need = -(-success // spell)
            left = 0
            right = len(potions) - 1
            answer = 0
            while left <= right:
                mid = left + (right - left) // 2
                if potions[mid] >= need:
                    right = mid - 1
                    answer = max(answer, len(potions) - mid)
                else:
                    left = mid + 1
            result.append(answer)
        return result